### PR TITLE
Add optional-iterator support to indexalator

### PR DESCRIPTION
**Pull request by davidwendt**
This PR adds support for the `optional-iterator` interface to the `cudf::detail::indexalator`. This will allow replacing usages of `pair-iterator` in future PR(s). This also includes gtests that were missing for the `indexalator`.

### CODE REVIEW

File: cpp/tests/iterator/indexalator_test.cu [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  auto d_col = cudf::test::fixed_width_column_wrapper<T>(
+    host_values.begin(), host_values.end(), validity.begin());
+
+  auto expected_values = thrust::host_vector<thrust::optional<cudf::size_type>>(host_values.size());
+
+  std::transform(host_values.begin(),
+                 host_values.end(),
+                 validity.begin(),
+                 expected_values.begin(),
+                 [host_values, validity](T v, bool b) {
```
Comment (vuule) [inline cpp/tests/iterator/indexalator_test.cu:88]:
```suggestion
                 [](T v, bool b) {
```

File: cpp/tests/iterator/indexalator_test.cu [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  auto d_col = cudf::test::fixed_width_column_wrapper<T>(
+    host_values.begin(), host_values.end(), validity.begin());
+
+  auto expected_values =
+    thrust::host_vector<thrust::pair<cudf::size_type, bool>>(host_values.size());
+  std::transform(host_values.begin(),
+                 host_values.end(),
+                 validity.begin(),
+                 expected_values.begin(),
+                 [host_values, validity](T v, bool b) {
```
Comment (vuule) [inline cpp/tests/iterator/indexalator_test.cu:64]:
```suggestion
                 [](T v, bool b) {
```

**Review by vuule — APPROVED** [review]
Looks good, just a few nitpicks.

File: cpp/include/cudf/detail/indexalator.cuh [hunk 3]
```diff
@@ -502,17 +502,32 @@ struct indexalator_factory {
...
-    nullable_index_accessor(scalar const& input) : has_nulls{!input.is_valid()}
+    scalar_nullable_index_accessor(scalar const& input) : is_null{!input.is_valid()}
     {
       iter = indexalator_factory::make_input_iterator(input);
     }
 
     __device__ thrust::pair<size_type, bool> operator()(size_type i) const
     {
-      return {iter[i], (has_nulls ? bit_is_set(null_mask, i + offset) : true)};
+      return {*iter, is_null};
```
Comment (ttnghia) [inline cpp/include/cudf/detail/indexalator.cuh:530]: I see that `nullable_index_accessor` above uses `iter[i]` while here is using `*iter`. We may need to make them consistent.

File: cpp/include/cudf/detail/indexalator.cuh [hunk 4]
```diff
@@ -530,7 +545,75 @@ struct indexalator_factory {
...
+      : null_mask{col.null_mask()}, offset{col.offset()}, has_nulls{has_nulls}
+    {
+      if (has_nulls) { CUDF_EXPECTS(col.nullable(), "Unexpected non-nullable column."); }
+      iter = make_input_iterator(col);
+    }
+
+    __device__ thrust::optional<size_type> operator()(size_type i) const
+    {
+      return has_nulls && !bit_is_set(null_mask, i + offset) ? thrust::nullopt
+                                                             : thrust::make_optional(iter[i]);
```
Comment (ttnghia) [inline cpp/include/cudf/detail/indexalator.cuh:575]: Similar to above: here is using `iter[i]` but below (line 598) uses `*iter`.

File: cpp/include/cudf/detail/indexalator.cuh (hunk 3 above)
Comment (davidwendt) [reply cpp/include/cudf/detail/indexalator.cuh:530]: This iterator is for a scalar which is only one value so using `iter[i]` would be invalid.

**Review by ttnghia — APPROVED** [review]
Just one more thing I'm a bit concerned but may not be able to address for now: The tests calls `this->iterator_test_thrust(expected_values, it_dev, host_values.size());`, instead of using the macro `EXPECT_COLUMNS_EQUAL` etc. If a test fails, we may have difficulty locating which line it is failing.

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**